go/libraries/doltcore/sqle/reflog_table_function.go

```diff
@@ -0,0 +1,202 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqle
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+	"github.com/dolthub/dolt/go/store/hash"
+)
+
+type ReflogTableFunction struct {
+	ctx      *sql.Context
+	database sql.Database
+	refExpr  sql.Expression
+}
+
+var _ sql.TableFunction = (*ReflogTableFunction)(nil)
+var _ sql.ExecSourceRel = (*ReflogTableFunction)(nil)
+
+var reflogTableSchema = sql.Schema{
+	&sql.Column{Name: "ref", Type: types.LongText},
+	&sql.Column{Name: "ref_timestamp", Type: types.Timestamp, Nullable: true},
+	&sql.Column{Name: "commit_hash", Type: types.LongText},
+	&sql.Column{Name: "commit_message", Type: types.LongText},
+}
+
+func (rltf *ReflogTableFunction) NewInstance(ctx *sql.Context, database sql.Database, expressions []sql.Expression) (sql.Node, error) {
+	newInstance := &ReflogTableFunction{
+		ctx:      ctx,
+		database: database,
+	}
+
+	node, err := newInstance.WithExpressions(expressions...)
+	if err != nil {
+		return nil, err
+	}
+
+	return node, nil
+}
+
+func (rltf *ReflogTableFunction) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	sqlDb, ok := rltf.database.(dsess.SqlDatabase)
+	if !ok {
+		return nil, fmt.Errorf("unexpected database type: %T", rltf.database)
+	}
+
+	target, err := rltf.refExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, fmt.Errorf("error evaluating expression (%s): %s",
+			rltf.refExpr.String(), err.Error())
+	}
+
+	refName, ok := target.(string)
+	if !ok {
+		return nil, fmt.Errorf("argument (%v) is not a string value, but a %T", target, target)
+	}
+
+	ddb := sqlDb.DbData().Ddb
+	journal := ddb.ChunkJournal()
+	if journal == nil {
+		return sql.RowsToRowIter(), nil
+	}
+
+	previousCommit := ""
+	rows := make([]sql.Row, 0)
+	err = journal.IterateRoots(func(root string, timestamp time.Time) error {
+		hashof := hash.Parse(root)
+		datasets, err := ddb.DatasetsByRootHash(ctx, hashof)
+		if err != nil {
+			return fmt.Errorf("unable to look up references for root hash %s: %s",
+				hashof.String(), err.Error())
+		}
+
+		return datasets.IterAll(ctx, func(id string, addr hash.Hash) error {
+			// If the caller has supplied a branch or tag name, without the fully qualified ref path,
+			// take the first match and use that as the canonical
+			if strings.HasSuffix(strings.ToLower(id), "/"+strings.ToLower(refName)) {
+				refName = id
```

Comment:
What happens if the name matches both a tag and a branch e.g.?

Reply:
Just matches the first one alphabetically?

Reply:
Yup, if the reflog contains entries for a tag and a branch with the same name, the first reflog entry that matches will be used for the fully qualified ref path. We iterate over the reflog from newest to oldest records, so the most recently used ref will match.